mava/wrappers/jaxmarl.py

```diff
@@ -271,22 +271,76 @@ def action_spec(self) -> specs.Spec:
         """Get action mask for each agent."""
-        if self.has_action_mask:
-            avail_actions = self._env.get_avail_actions(state)
-            mask = jnp.array(batchify(avail_actions, self.agents), dtype=jnp.float32)
-        else:
-            mask = jnp.ones(self._action_shape, dtype=jnp.float32)
-        return mask
+        ...
+
+    @abstractmethod
+    def get_global_state(self, env_state: Array, obs: Dict[str, Array]) -> Array:
+        """Get global state from observation for each agent."""
+        ...
+
+    @property
+    @abstractmethod
+    def state_size(self) -> chex.Array:
+        ...
+
+
+class SmaxWrapper(JaxMarlWrapper):
+    """WWrapper for SMAX environment"""
+
+    def __init__(self, env: MultiAgentEnv, has_global_state: bool = False, timelimit: int = 500):
+        super().__init__(env, has_global_state, timelimit)
+
+    def state_size(self) -> chex.Array:
+        return self._env.state_size
 
-    def get_global_state(self, obs: Dict[str, Array]) -> Array:
+    def action_mask(self, state: JaxMarlState) -> Array:
+        """Get action mask for each agent."""
+        avail_actions = self._env.get_avail_actions(state)
+        return jnp.array(batchify(avail_actions, self.agents), dtype=jnp.float32)
+
+    def get_global_state(self, env_state: Array, obs: Dict[str, Array]) -> Array:
         """Get global state from observation and copy it for each agent."""
-        return jnp.tile(jnp.array(obs["world_state"]), (self._env.num_agents, 1))
+        return jnp.tile(jnp.array(obs["world_state"]), (self.num_agents, 1))
+
+
+class MabraxWrapper(JaxMarlWrapper):
+    """Wrraper for the Mabrax environment."""
+
+    def __init__(self, env: MABraxEnv, has_global_state: bool = False, timelimit: int = 1000):
+        super().__init__(env, has_global_state, timelimit)
+        self.action_shape = self.action_spec().shape
+
+    def state_size(self) -> chex.Array:
+        state_size = self._env.env.observation_size
+        return (
+            state_size
+            if self._env.homogenisation_method != "max"
+            else state_size + self._env.num_agents
+        )
+
+    def action_mask(self, state: JaxMarlState) -> Array:
+        """Get action mask for each agent."""
+        return jnp.ones(self.action_shape, dtype=jnp.float32)
+
+    def get_global_state(self, env_state: BraxState, obs: Dict[str, Array]) -> Array:
+        """Get global state from observation and copy it for each agent."""
+        # Use the global state of brax.
+        global_state = jnp.tile(env_state.obs, (self.num_agents, 1))
+
+        #  In this case, add_agent_id=False so the agent's ID must be added to the global state.
+        if self._env.homogenisation_method == "max":
+            agent_ids = jnp.eye(self.num_agents)
+            global_state = jnp.tile(env_state.obs, (self.num_agents, 1))
+            global_state = jnp.concatenate([agent_ids, global_state], axis=-1)
```

Comment:
I don't think we ever need to add agent_ids to global state as all critics would get the same state, so no need to differentiate?

Also @RuanJohn @OmaymaMahjoub do we want' multiple copies of the global state per agent? To me it seems like something you'd only pass this to one critic, at least this is how I'm doing it in MA-SAC :thinking:

Reply:
Ok chatted to @RuanJohn about this and it seems like it was done for convenience and it would be a mission to change it in the PPO systems, so I'll work around this for SAC.

As for adding agent ids to the global state, can we comment here that we've found that having IDs in the global state is usually good, but can you also add an option for this so it's easier to experiment with? `if self._env.homogenisation_method == "max" and add_agent_ids:`

Reply:
Done ✅